src/ripple/nodestore/impl/TaskQueue.h

```diff
@@ -0,0 +1,58 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_TASKQUEUE_H_INCLUDED
+#define RIPPLE_NODESTORE_TASKQUEUE_H_INCLUDED
+
+#include <ripple/core/impl/Workers.h>
+
+#include <functional>
+#include <queue>
+
+namespace ripple {
+namespace NodeStore {
+
+class TaskQueue : private Workers::Callback
+{
+public:
+    TaskQueue();
+
+    /** Adds a task to the queue
+
+        @param task std::function with signature void()
+    */
+    void
+    addTask(std::function<void()> task);
+
+    void
+    pause() {workers_.pauseAllThreadsAndWait();}
```

Comment:
At some point, I did encounter an issue where an outstanding task that held a shared shard pointer would cause the NuDB context object to throw in the DatabaseShardImp dtor. This occurred because the task shard pointer kept a NuDB backend open and the context object expected them all to be closed at destruction. Calling `pause()` in `onStop` fixes the issue as it effectively stops all tasks and awaits any ongoing to complete. If I ever come across a related shutdown problem again, I'll look into `ClosureCounter` and ask you for details on your technique.

Reply:
Since you've seen a failing case that is hard to reproduce, let me tell you about the technique and you can decide whether it's applicable.

The basic idea is that you have two shell scripts running in different terminals:

1. One shell script runs a loop that sleeps for 5 seconds and then starts rippled.  If rippled exits with a code of zero it sleeps 5 seconds and restarts rippled.  Repeat ad infinitum.

2. The second shell script also runs a (slightly more complicated) loop.
    1. Wait until rippled is running (I use pgrep).
    1. Wait until the server_state reaches full and remains at full for some duration.
    1. After rippled has been stable at full for long enough send "rippled stop"
    1. Wait until rippled actually does stop.
    1. Return to the top of the loop.

I occasionally tune the second script to look for some condition in addition to rippled running for a certain amount of time.  There's flexibility.

If you set both of these scripts in motion you'll have rippled gracefully starting and stopping in a loop.  The loop stops if rippled hangs or exits with a code other than zero.  Let that run for a few days.

If that sounds like it might be useful I can send you the two scripts I'm using.  Then you can modify them for your own purposes.

Reply:
@scottschurr I just pushed a commit addressing your feedback. Please let me know if I missed something.

The scripts sound like a very useful test, please send them to me. I'll run them for a few days.